### PR TITLE
doc: improve doc/dev/encoding.rst

### PR DESCRIPTION
**Pull request by rzarzynski**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [x] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: doc/dev/encoding.rst [hunk 1]
```diff
@@ -3,9 +3,74 @@ Serialization (encode/decode)
...
-``encode`` and ``decode`` methods that write and read from ``bufferlist``
-objects representing byte strings.
+encoded into a string of bytes. Usually (but not always -- multiple
+serialization facilities coexist in Ceph) serializable structures
+have ``encode`` and ``decode`` methods that write and read from
+``bufferlist`` objects representing byte strings.
+
+Terminology
+-----------
+It is useful to think not in the domain of daemons and clients but
```
Comment (anthonyeleven) [inline doc/dev/encoding.rst:13]: s/useful/best/

File: doc/dev/encoding.rst [hunk 2]
```diff
@@ -3,9 +3,74 @@ Serialization (encode/decode)
...
+encoders and decoders. An encoder serializes a structure into bufferlist
```
Comment (anthonyeleven) [inline doc/dev/encoding.rst:14]: into a bufferlist
while a decoder

File: doc/dev/encoding.rst [hunk 3]
```diff
@@ -3,9 +3,74 @@ Serialization (encode/decode)
...
+
+Principles for format change
+----------------------------
+It is not unusual that the format of serialization changes. This
+process requires careful attention from during both development
+and review.
+
+The general rule is that a decoder must understand what had been
+encoded by an encoder. Most of the problems come from ensuring
+this stays also between old decoders and new encoders as well as
```
Comment (anthonyeleven) [inline doc/dev/encoding.rst:39]: ensuring that compatibility continues between old

File: doc/dev/encoding.rst [hunk 4]
```diff
@@ -3,9 +3,74 @@ Serialization (encode/decode)
...
+new decoders and old decoders. It is because one shall assume
```
Comment (anthonyeleven) [inline doc/dev/encoding.rst:40]: decoders.  One should assume

File: doc/dev/encoding.rst [hunk 5]
```diff
@@ -93,3 +158,69 @@ because we might still be passed older-versioned messages that do not
 have the field.  The ``struct_v`` variable is a local set by the ``DECODE_START``
 macro.
 
+# Into the weeeds
+
+The append-extendability of our dencoders is a result of the forward
+compatibility the ``ENCODE_START`` and ``DECODE_FINISH`` macros bring.
```
Comment (anthonyeleven) [inline doc/dev/encoding.rst:164]: compatibility that the

File: doc/dev/encoding.rst [hunk 6]
```diff
@@ -93,3 +158,69 @@ because we might still be passed older-versioned messages that do not
...
+
+They are implementing extendibility facilities. Encoder, when filling
```
Comment (anthonyeleven) [inline doc/dev/encoding.rst:166]: They implement
An encoder, when filling

**Review by anthonyeleven — APPROVED** [review]
Several fussy changes suggested

File: doc/dev/encoding.rst [hunk 7]
```diff
@@ -3,9 +3,74 @@ Serialization (encode/decode)
...
+
+With this being said, there are few rules limiting the degree
+of interoperability between dencoders:
+
+* ``n-2`` for dencoding between daemons,
+* ``n-3`` hard requirement for client-involved scenarios,
+* ``n-3..``  soft requirements for clinet-involved scenarios. Ideally
+  every client should be able to talk any version of daemons.
+
+As the underlying reasons are the same, the rules dencoders
```
Comment (rzarzynski) [inline doc/dev/encoding.rst:61]: This depends on https://github.com/ceph/ceph/pull/51488#pullrequestreview-1437514892.

File: doc/dev/encoding.rst [hunk 8]
```diff
@@ -3,9 +3,74 @@ Serialization (encode/decode)
...
+   from Ceph upgrades. Moreover, proliferation of containerization
+   bring the variability even to e.g. ``librbd`` -- now user space
+   libraries live on the container own.
+
+With this being said, there are few rules limiting the degree
+of interoperability between dencoders:
+
+* ``n-2`` for dencoding between daemons,
+* ``n-3`` hard requirement for client-involved scenarios,
+* ``n-3..``  soft requirements for clinet-involved scenarios. Ideally
```
Comment (neha-ojha) [inline doc/dev/encoding.rst:58]: nit: client-involved

File: doc/dev/encoding.rst [hunk 9]
```diff
@@ -93,3 +158,69 @@ because we might still be passed older-versioned messages that do not
...
+          decode(struct_len, bl);					\
+          ...                                                           \
+          struct_end = bl.get_off() + struct_len;			\
+          }								\
+          do {
+
+
+Decoder uses this information to discard the extra bytes it does not
+understand. Advancing bufferlist is critical as dencoders tend to be nested;
+just leaving it intact would work only for the very last ``deocde`` call
```
Comment (neha-ojha) [inline doc/dev/encoding.rst:210]: nit: decode

File: doc/dev/encoding.rst [hunk 10]
```diff
@@ -93,3 +158,69 @@ because we might still be passed older-versioned messages that do not
...
+          ceph_le32 struct_len;                                         \
+          auto filler = (bl).append_hole(sizeof(struct_v) +             \
+            sizeof(struct_compat) + sizeof(struct_len));                \
+          const auto starting_bl_len = (bl).length();                   \
+          using ::ceph::encode;                                         \
+          do {
+
+The ``struct_len`` field allows the decoder to eat all the bytes that were
+left undecoded in the user-provided ``decode`` implementation.
+Analogically, decoders tracks how much input has been decoded in the
```
Comment (neha-ojha) [inline doc/dev/encoding.rst:193]: nit: track

**Review by neha-ojha — APPROVED** [review]

File: doc/dev/encoding.rst (hunk 7 above)
Comment (idryomov) [reply doc/dev/encoding.rst:61]: I commented in https://github.com/ceph/ceph/pull/51488 -- I think all mention of `n-3` for client-involved cases must go.

File: doc/dev/encoding.rst [hunk 11]
```diff
@@ -3,9 +3,74 @@ Serialization (encode/decode)
...
+possible in a cluster. There are 2 main reasons for that:
+
+1. Upgrades. Although there are recommendations related to the order
+   of entity types (mons/osds/clients), it is not mandatory and
+   no assumption should be made about it.
+2. Huge variability of client versions. It was always the case
+   that kernel (and thus kernel clients) upgrades are decoupled
+   from Ceph upgrades. Moreover, proliferation of containerization
+   bring the variability even to e.g. ``librbd`` -- now user space
+   libraries live on the container own.
```
Comment (idryomov) [inline doc/dev/encoding.rst:51]:
```suggestion
   libraries can live in their own container.
```

File: doc/dev/encoding.rst [hunk 12]
```diff
@@ -3,9 +3,74 @@ Serialization (encode/decode)
...
+these roles can swing -- just imagine handling of the response: OSD encodes
+the ``MOSDOpReply`` while RBD clients decode.
+
+Encoder and decoder operate accordingly to a format which is defined
+by a programmer by implementing the ``encode`` and ``decode`` methods.
+
+Principles for format change
+----------------------------
+It is not unusual that the format of serialization changes. This
+process requires careful attention from during both development
```
Comment (idryomov) [inline doc/dev/encoding.rst:34]:
```suggestion
process requires careful attention during both development
```

File: doc/dev/encoding.rst [hunk 13]
```diff
@@ -3,9 +3,74 @@ Serialization (encode/decode)
...
+that compatibility continues between old decoders and new encoders
+as well as new decoders and old decoders. One should assume
+that -- if not otherwise derogated -- any mix (old/new) is
+possible in a cluster. There are 2 main reasons for that:
+
+1. Upgrades. Although there are recommendations related to the order
+   of entity types (mons/osds/clients), it is not mandatory and
+   no assumption should be made about it.
+2. Huge variability of client versions. It was always the case
+   that kernel (and thus kernel clients) upgrades are decoupled
```
Comment (idryomov) [inline doc/dev/encoding.rst:48]: "The kernel (and thus kernel client) upgrades have always been decoupled from Ceph upgrades."

File: doc/dev/encoding.rst [hunk 14]
```diff
@@ -3,9 +3,74 @@ Serialization (encode/decode)
...
+   from Ceph upgrades. Moreover, proliferation of containerization
+   bring the variability even to e.g. ``librbd`` -- now user space
```
Comment (idryomov) [inline doc/dev/encoding.rst:50]:
```suggestion
   brought the variability even to e.g. ``librbd`` -- now user space
```

File: doc/dev/encoding.rst [hunk 15]
```diff
@@ -3,9 +3,74 @@ Serialization (encode/decode)
...
+   libraries live on the container own.
+
+With this being said, there are few rules limiting the degree
+of interoperability between dencoders:
+
+* ``n-2`` for dencoding between daemons,
+* ``n-3`` hard requirement for client-involved scenarios,
```
Comment (jdurgin) [inline doc/dev/encoding.rst:57]: to be explicit, n-3 is what we test

**Comment by zdover23** [comment]
I'm merging this PR. A follow-up PR will incorporate all notes added to this PR.

**Comment by zdover23** [comment]
https://github.com/ceph/ceph/pull/52758 - Reef backport
https://github.com/ceph/ceph/pull/52759 - Quincy backport

**Pull request merged**